### PR TITLE
feat(dashboard,medusa): Promotion Campaign fixes

### PR DESCRIPTION
**Pull request by riqwan**
what: 

- introduces currency code to promotions
- adds validations based on currency
- scope currencies by store currencies
- scope promotion add to campaign with currency
- scope campaigns with correct currency
- promotion currency carried over to campaign create
- use select instead of combobox for equals operator
- resolve type issues
- a whole load of fixes

### CODE REVIEW

File: packages/modules/promotion/src/models/campaign.ts [hunk 1]
```diff
@@ -40,8 +39,8 @@ export default class Campaign {
   @Property({ columnType: "text", nullable: true })
   description: string | null = null
 
-  @Property({ columnType: "text", nullable: true })
-  currency: string | null = null
+  @Property({ columnType: "text" })
+  currency: string
```
Comment (olivermrbl) [inline packages/modules/promotion/src/models/campaign.ts]: question: If the campaign has a spend-budget, I would imagine the currency could be nullable. What prompted this change?

File: packages/modules/promotion/src/services/promotion-module.ts [hunk 2]
```diff
@@ -475,6 +476,11 @@ export default class PromotionModuleService<
     const promotionsData: CreatePromotionDTO[] = []
     const applicationMethodsData: CreateApplicationMethodDTO[] = []
     const campaignsData: CreateCampaignDTO[] = []
+    const existingCampaigns = await this.campaignService_.list(
+      { id: data.map((d) => d.campaign_id).filter((id) => isString(id)) },
```
Comment (olivermrbl) [inline packages/modules/promotion/src/services/promotion-module.ts:481]: thought: Could we do:
```suggestion
      { id: data.filter((d) => isPresent(d.campaign_id)) },
```

File: packages/modules/promotion/src/services/promotion-module.ts [hunk 3]
```diff
@@ -522,6 +526,37 @@ export default class PromotionModuleService<
...
+        existingCampaigns.find((c) => c.id === campaignId)?.currency
+
+      if (
+        !campaignCurrency ||
+        campaignCurrency !== applicationMethodData?.currency_code
+      ) {
+        throw new MedusaError(
+          MedusaError.Types.INVALID_DATA,
+          `Currency between promotion and campaigns should match`
+        )
```
Comment (olivermrbl) [inline packages/modules/promotion/src/services/promotion-module.ts:559]: Goes along with my question on the data model. If the campaign is of type spend, do we need to enforce this match.

File: packages/modules/promotion/src/services/promotion-module.ts [hunk 4]
```diff
@@ -721,20 +759,36 @@ export default class PromotionModuleService<
...
+          `Could not find campaign with id ${campaignId}`
+        )
+      }
+
+      if (campaignId && existingCampaign?.currency !== promotionCurrencyCode) {
+        throw new MedusaError(
+          MedusaError.Types.INVALID_DATA,
+          `Currency code doesn't match for campaign (${campaignId}) and promotion (${existingPromotion.id})`
+        )
+      }
```
Comment (olivermrbl) [inline packages/modules/promotion/src/services/promotion-module.ts]: Same as other question.

File: packages/modules/promotion/src/services/promotion-module.ts [hunk 5]
```diff
@@ -1292,7 +1348,19 @@ export default class PromotionModuleService<
...
+    const promotionsWithInvalidCurrency = promotionsToAdd.filter(
+      (promotion) =>
+        promotion.application_method?.currency_code !== campaign.currency
+    )
+
+    if (promotionsWithInvalidCurrency.length > 0) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        `Cannot add promotions to campaign where currency_code don't match.`
+      )
```
Comment (olivermrbl) [inline packages/modules/promotion/src/services/promotion-module.ts:1414]: Same as other question.

**Review by olivermrbl — COMMENTED** [review]
Could I get you to share a bit about the background for adding currencies to promotions and campaigns in this PR? I think that would make it easier to review.

File: packages/admin-next/dashboard/src/v2-routes/promotions/promotion-create/components/create-promotion-form/create-promotion-form.tsx [hunk 6]
```diff
@@ -146,53 +147,54 @@ export const CreatePromotionForm = ({
...
+          target_rules: buildRulesData(targetRulesData),
+          buy_rules: buildRulesData(buyRulesData),
         },
         is_automatic: is_automatic === "true",
       }).then(() => handleSuccess())
     },
     async (error) => {
       // TODO: showcase error when something goes wrong
       // Wait for alert component and use it here
+      console.log("error - ", error)
```
Comment (riqwan) [inline packages/admin-next/dashboard/src/v2-routes/promotions/promotion-create/components/create-promotion-form/create-promotion-form.tsx]:
```suggestion
```

File: packages/admin-next/dashboard/src/v2-routes/promotions/promotion-create/components/create-promotion-form/create-promotion-form.tsx [hunk 7]
```diff
@@ -316,11 +334,37 @@ export const CreatePromotionForm = ({
...
+    (rule) => rule.attribute === "currency_code"
+  )
+
+  if (watchCurrencyRule) {
+    const formData = form.getValues()
+    const currencyCode = formData.application_method.currency_code
+    const ruleValue = watchCurrencyRule.values
+
+    if (!Array.isArray(ruleValue) && currencyCode !== ruleValue) {
+      console.log("here?")
```
Comment (riqwan) [inline packages/admin-next/dashboard/src/v2-routes/promotions/promotion-create/components/create-promotion-form/create-promotion-form.tsx]:
```suggestion
```

File: packages/admin-next/dashboard/src/v2-routes/promotions/promotion-add-campaign/promotion-add-campaign.tsx [hunk 8]
```diff
@@ -11,13 +11,24 @@ export const PromotionAddCampaign = () => {
...
+
   const {
     campaigns,
     isPending: areCampaignsLoading,
     isError: isCampaignError,
     error: campaignError,
-  } = useCampaigns()
-
+  } = useCampaigns(campaignQuery)
+  console.log("campaigns --- ", campaigns)
```
Comment (riqwan) [inline packages/admin-next/dashboard/src/v2-routes/promotions/promotion-add-campaign/promotion-add-campaign.tsx]:
```suggestion
```

File: packages/admin-next/dashboard/src/v2-routes/campaigns/common/components/create-campaign-form-fields/create-campaign-form-fields.tsx [hunk 9]
```diff
@@ -98,33 +125,7 @@ export const CreateCampaignFormFields = ({ form, fieldScope = "" }) => {
...
-                      ))}
-                    </Select.Content>
-                  </Select>
-                </Form.Control>
-                <Form.ErrorMessage />
-              </Form.Item>
-            )
-          }}
-        />
+        <div></div>
```
Comment (olivermrbl) [inline packages/admin-next/dashboard/src/v2-routes/campaigns/common/components/create-campaign-form-fields/create-campaign-form-fields.tsx]: rm?
```suggestion
```

File: packages/core/types/src/promotion/http.ts [hunk 10]
```diff
@@ -4,6 +4,7 @@ export type AdminGetPromotionRulesRes = {
   attribute_label: string
   field_type?: string
   operator: string
+  hydrate: boolean
```
Comment (olivermrbl) [inline packages/core/types/src/promotion/http.ts:7]: thought: Could I get you to write a brief doc on how all of these properties tie together and their function, specifically hydrate and disguised? I lost track of that over the past months 😅

File: packages/medusa/src/api/admin/promotions/utils/rule-attributes-map.ts [hunk 11]
```diff
@@ -1,6 +1,7 @@
 export enum DisguisedRule {
```
Comment (olivermrbl) [inline packages/medusa/src/api/admin/promotions/utils/rule-attributes-map.ts:1]: Same for this. Please include this attributes map in the doc 🙏

File: packages/medusa/src/api/admin/promotions/utils/rule-query-configuration.ts [hunk 12]
```diff
@@ -4,9 +4,9 @@ export const ruleQueryConfigurations = {
     labelAttr: "name",
```
Comment (olivermrbl) [inline packages/medusa/src/api/admin/promotions/utils/rule-query-configuration.ts:4]: And this :D

**Review by olivermrbl — APPROVED** [review]
I think this is a bit too hard to review as it is composed of many small changes. Therefore, I would rather just have this merged and then review the promotions module all together in the next couple of months.

What do you think? 

That said, I left a few comments on things that would be good to have documented for understanding.

File: packages/admin-next/dashboard/src/v2-routes/campaigns/common/components/create-campaign-form-fields/create-campaign-form-fields.tsx (hunk 9 above)
Comment (riqwan) [reply packages/admin-next/dashboard/src/v2-routes/campaigns/common/components/create-campaign-form-fields/create-campaign-form-fields.tsx]: This is actually needed to keep the grid spot the same size :D

File: packages/core/types/src/promotion/http.ts (hunk 10 above)
Comment (riqwan) [reply packages/core/types/src/promotion/http.ts:7]: Yes will do, I think most of these can now be removed as we don't use them anymore. Will do a cleanup after all the kinks are sorted. 🤞🏻

Comment (olivermrbl) [reply packages/core/types/src/promotion/http.ts:7]: Great 💪

File: packages/admin-next/dashboard/src/v2-routes/campaigns/common/components/create-campaign-form-fields/create-campaign-form-fields.tsx (hunk 9 above)
Comment (olivermrbl) [reply packages/admin-next/dashboard/src/v2-routes/campaigns/common/components/create-campaign-form-fields/create-campaign-form-fields.tsx]: Ahh I see :smile

**Pull request merged**